src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs

```diff
@@ -0,0 +1,238 @@
+        /// <exception cref="ArgumentNullException">The parameter 'token' is null.</exception>
+        /// <exception cref="ArgumentException">The token is not assignable from <see cref="Saml2SecurityToken"/>.</exception>
+        /// <exception cref="InvalidOperationException">Configuration <see cref="SecurityTokenHandlerConfiguration"/>is null.</exception>
+        /// <exception cref="ArgumentException">Saml2SecurityToken.Assertion is null.</exception>
+        /// <exception cref="SecurityTokenValidationException">Thrown if Saml2SecurityToken.Assertion.SigningToken is null.</exception>
+        /// <exception cref="SecurityTokenValidationException">Thrown if the certificate associated with the token issuer does not pass validation.</exception>
+        public override ReadOnlyCollection<ClaimsIdentity> ValidateToken(SecurityToken token)
+        {
+            Saml2SecurityToken samlToken = (Saml2SecurityToken)token;
+            string assertionXML = samlToken.AssertionXML;
+            SamlTokenValidationParameters tokenValidation = new SamlTokenValidationParameters();
+            MSIdentityTokens.SecurityToken msIdentityToken;
+            ClaimsPrincipal claim = _internalSaml2SecurityTokenHandler.ValidateToken(assertionXML,
+                tokenValidation.ConvertToTokenValidationParameters(Configuration, samlToken.InternalSaml2SecurityToken), out msIdentityToken);
+            return claim.Identities?.ToList().AsReadOnly();
+        }
+        #endregion
+
+        #region TokenSerialization
+
+        /// <summary>
+        /// Indicates whether the current XML element can be read as a token 
+        /// of the type handled by this instance.
+        /// </summary>
+        /// <param name="reader">An XML reader positioned at a start 
+        /// element. The reader should not be advanced.</param>
+        /// <returns>'True' if the ReadToken method can the element.</returns>
+        public override bool CanReadToken(XmlReader reader)
+        {
+            return _internalSaml2SecurityTokenHandler.CanReadToken(reader);
+        }
+
+        /// <summary>
+        /// Deserializes from XML a token of the type handled by this instance.
+        /// </summary>
+        /// <param name="reader">An XML reader positioned at the token's start 
+        /// element.</param>
+        /// <returns>An instance of <see cref="Saml2SecurityToken"/>.</returns>
+        /// <exception cref="InvalidOperationException">Is thrown if 'Configuration' or 'Configruation.IssuerTokenResolver' is null.</exception>
+        public override SecurityToken ReadToken(XmlReader reader)
+        {
+            XmlDocument doc = new XmlDocument();
+            XmlElement rstXml = (doc.ReadNode(reader) as XmlElement);
+            MSIdentityTokens.Saml2.Saml2SecurityToken internalSecurityToken = (MSIdentityTokens.Saml2.Saml2SecurityToken)_internalSaml2SecurityTokenHandler.ReadToken(rstXml.OuterXml);
```

Comment:
There's no need to go via an XmlElement to a string, the IM token handler has an overload that accepts an XmlReader. I had a look at the implementation of the overload you are using and it actually creates an XmlReader on the passed in string to then read the token. So you're going from XmlReader -> XmlElement -> String -> XmlReader. There's also a method called ReadSamlToken which returns the type you are casting to so you can skip the case by using the correct method.

Reply:
we have to keep the assertionxml so that during validate we can use to validate. The ValidateToken doesn't accept the token, else there was no need to keep the assertionxml.

Reply:
As discussed I am going to submit a PR to other repo and fix this performance problem once the other issues is submitted and released.

Reply:
https://github.com/CoreWCF/CoreWCF/issues/579